Added gltf OMI_collider extension support.

cyberneticocult opened the pull request:
Added support for the OMI_collider extension proposed by the OMI group in the following specification.
https://github.com/cyberneticocult/gltf-extensions-omi-collider/tree/main/extensions/2.0/OMI_collider

Scenes can be exported with an extension addon for the blender gltf exporter.
https://github.com/cyberneticocult/gltf-blender-io-omi-collision-extension

Test data for scenes that use the OMI_collider extension can be found at the following static resource urls.
https://cyberneticocult.github.io/webaverse-omi-collider-examples/boxes-glb/
https://cyberneticocult.github.io/webaverse-omi-collider-examples/boxes-gltf/boxes.gltf
https://cyberneticocult.github.io/webaverse-omi-collider-examples/collider-types-glb/
https://cyberneticocult.github.io/webaverse-omi-collider-examples/collider-types-gltf/collider_types.gltf

cyberneticocult committed:
Added OMI_collider support to loading of .glb files when building physics.

cyberneticocult commented:
I should probably mention that I didn't add the "isTrigger" part of the specification. I wasn't sure how you would want to handle this in the client.

I'm guessing there's a way of registering a collision hull in physx that does an intersection test with other bodies but won't collide with them? Also, would we emit those events on the app the node is a child of? Should the user of the app register listeners for the things they want to detect, or do we just emit all intersection events?

memelotsqui commented:
Ive been doing some testings and so far everything seems correct, but im just double checking since some of the exported examples from  https://github.com/cyberneticocult/webaverse-omi-collider-examples, are not working correctly (negative-scale-gltf is not working correctly), but may only be to certain specific cases, as the code looks correct to me.

Collider-types example is working correctly: https://github.com/cyberneticocult/webaverse-omi-collider-examples/tree/main/collider-types-glb So I want to do another test with a custom omi export just to double check everything is correct. (this pr even considered the capsule has to be rotated to adapt to physx as shown in the image)

![image](https://user-images.githubusercontent.com/1117257/191613512-2260bf38-6093-41ee-b0a1-3b6ea3691e0c.png)


https://user-images.githubusercontent.com/1117257/191613110-0d015915-7d5b-4d7d-93cb-bb3dbde5dfd5.mp4

cyberneticocult commented:
I didn't actually test out the negative scale data. I wasn't sure how this should be handled. I'm guessing everything should just use absolute values, but I never checked to see how physx would handle negative scale values.

cyberneticocult commented:
I also left compound colliders undefined, but after some discussions in the OMI discord channel for extensions I was able to get a better idea of what they are.

From the sounds of it, compound colliders are meant to be the same as what Godot has for compound shapes.
https://docs.godotengine.org/en/stable/tutorials/assets_pipeline/escn_exporter/physics.html?highlight=compound#collision-shapes

Colliders can be parented to a compound collider which they will inherit their properties from. I think for the OMI_collider spec the only thing that would be inherited is the "isTrigger" property, but when combined with the OMI_physics_body spec, this will probably have more use.

I was wondering for compounds, when building them, should they all be joined into a single mesh? That would seem to lose the benefit of using box, capsule and hull colliders as they should be cheaper to test against right?

cyberneticocult committed:
Added console warnings for negative scale and compound type nodes.

cyberneticocult commented:
I've added some warnings to handle nodes with negative scale or that have been set to compound type. Hopefully that should do for now. There's another related OMI spec (OMI_physics_body) that when it's further along I think it'll make sense to revisit the compound collider to make building dynamic colliders easier.

cyberneticocult commented:
Just also wanted to mention I've added some additional features to the blender addon for making exporting colliders easier.
https://github.com/cyberneticocult/gltf-blender-io-omi-collision-extension/commit/0c7006b5d730191795ea4a4c0da60f2ea96d03b8

I've also updated a scene I've been working on with OMI colliders that could be helpful for providing test data.
https://cyberneticocult.github.io/webaverse-ld-block-kits/ld-block-kit-01/

lalalune commented:
@memelotsqui can you migrate this to https://github.com/webaverse/compiler please :)

memelotsqui commented:
Alright!, will get to work on it :)
